File: Day248_2025/Q248_GFG.py
class Solution:
    def segregate(self, head):
        if not head:
            return head

        cntZero = cntOne = cntTwo = 0
        curr = head

        # Count 0s, 1s, 2s
        while curr:
            if curr.data == 0:
                cntZero += 1
            elif curr.data == 1:
                cntOne += 1
            else:
                cntTwo += 1
            curr = curr.next

        # Rewrite values in sorted order
        curr = head
        while curr:
            if cntZero > 0:
                curr.data = 0
                cntZero -= 1
            elif cntOne > 0:
                curr.data = 1
                cntOne -= 1
            else:
                curr.data = 2
                cntTwo -= 1
            curr = curr.next

        return head

File: Day248_2025/Q248_GFG.java
class Q248_GFG {
    public Node segregate(Node head) {
        if (head == null) return head;

        Node curr = head;
        int cntZero = 0, cntOne = 0, cntTwo = 0;

        // Count occurrences of 0, 1, and 2
        while (curr != null) {
            if (curr.data == 0) cntZero++;
            else if (curr.data == 1) cntOne++;
            else cntTwo++;
            curr = curr.next;
        }

        // Rewrite linked list with sorted values
        curr = head;
        while (curr != null) {
            if (cntZero > 0) {
                curr.data = 0;
                cntZero--;
            } else if (cntOne > 0) {
                curr.data = 1;
                cntOne--;
            } else {
                curr.data = 2;
                cntTwo--;
            }
            curr = curr.next;
        }

        return head;
    }
}
